### PR TITLE
Product Management

### PR DESCRIPTION
**Pull request by Domarkaite**
Added product management (Products and Product Variations). As of right now, the Products and Product Variations are in different controllers. There are some status codes missing because of validation

### CODE REVIEW

File: Shopipy.BusinessManagement/Properties/launchSettings.json
Comment (imbieras) [inline Shopipy.BusinessManagement/Properties/launchSettings.json:1]: Why is this here, when launching is done for the whole API through Aspire or Docker?

File: Shopipy.CategoryManagement/Properties/launchSettings.json
Comment (imbieras) [inline Shopipy.CategoryManagement/Properties/launchSettings.json:1]: Why is this here, when launching is done for the whole API through Aspire or Docker?

File: Shopipy.Persistence/Properties/launchSettings.json
Comment (imbieras) [inline Shopipy.Persistence/Properties/launchSettings.json:1]: Why is this here, when launching is done for the whole API through Aspire or Docker?

File: Shopipy.ProductManagement/Controllers/ProductController.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Shopipy.Persistence.Models;
+using Shopipy.ProductManagement.DTOs;
+using Shopipy.Shared;
+using Shopipy.Shared.Services;
+
+namespace Shopipy.ProductManagement.Controllers
+{
+    [Route("{businessId}/products")]
```
Comment (imbieras) [inline Shopipy.ProductManagement/Controllers/ProductController.cs:11]: https://github.com/imbieras/shopipy/blob/0523318744943e2964beba1a74ff2fd0e963b0b3/Shopipy.CategoryManagement/Controllers/CategoryController.cs#L11 or any other Management. Basically this doesn't follow the convention.

File: Shopipy.ProductManagement/Controllers/ProductController.cs
Comment (imbieras) [inline Shopipy.ProductManagement/Controllers/ProductController.cs:1]: The original swagger doc that we got was missing a GET method to get a list of products a business offers, we should probably add that :)

File: Shopipy.BusinessManagement/Properties/launchSettings.json
Comment (Buztas) [inline Shopipy.BusinessManagement/Properties/launchSettings.json:1]: Remove launchSettings.json files

File: Shopipy.CategoryManagement/Properties/launchSettings.json
Comment (Buztas) [inline Shopipy.CategoryManagement/Properties/launchSettings.json:1]: Same here

File: Shopipy.ProductManagement/DependencyInjection.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿using Shopipy.Shared.Services;
+using Shopipy.ProductManagement.Services;
+
+namespace Shopipy.ProductManagement
+{
+    public static class DependencyInjection
+    {
+        public static IServiceCollection AddCategoryManagement(this IServiceCollection categories)
```
Comment (Buztas) [inline Shopipy.ProductManagement/DependencyInjection.cs:8]: Probably forgot to change, but should be AddProductManagement

File: Shopipy.ProductManagement/Mappings/ProductMappingProfile.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        public ProductMappingProfile()
+        {
+            CreateMap<Product, ProductResponseDTO>();
+
+            CreateMap<ProductRequestDTO, Product>()
+                .ForMember(dest => dest.BusinessId, opt => opt.Ignore()) 
+                .ForMember(dest => dest.ProductId, opt => opt.Ignore()) 
+                .ForMember(dest => dest.CreatedAt, opt => opt.Ignore()) 
+                .ForMember(dest => dest.UpdatedAt, opt => opt.Ignore())
+                .ForMember(dest => dest.CategoryId, opt => opt.MapFrom(src => src.CategoryId))
```
Comment (Buztas) [inline Shopipy.ProductManagement/Mappings/ProductMappingProfile.cs:18]: As konstantinas previously mentioned, whenever you are using src => src.sameNameAsDest variable, you dont need these lines, because automapper does it for you.

File: Shopipy.ProductManagement/Services/ProductService.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
+﻿using Shopipy.Persistence.Models;
+using Shopipy.Persistence.Repositories;
+using Shopipy.Shared.Services;
+
+namespace Shopipy.ProductManagement.Services
+{
+    public class ProductService : IProductService
```
Comment (Buztas) [inline Shopipy.ProductManagement/Services/ProductService.cs:7]: You could use primary constructors here, Konstantinas might be very angry about not using them

File: Shopipy.ProductManagement/Services/ProductVariationService.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
+﻿using Shopipy.Persistence.Models;
+using Shopipy.Persistence.Repositories;
+using Shopipy.Shared.Services;
+
+namespace Shopipy.ProductManagement.Services
+{
+    public class ProductVariationService : IProductVariationService
```
Comment (Buztas) [inline Shopipy.ProductManagement/Services/ProductVariationService.cs:7]: Same here

File: Shopipy.ProductManagement/Services/ProductService.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        private readonly IGenericRepository<Product> _productRepository;
+        private readonly IGenericRepository<ProductVariation> _variationRepository;
```
Comment (imbieras) [inline Shopipy.ProductManagement/Services/ProductService.cs]: Not used?

File: Shopipy.Shared/Services/IProductService.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using Shopipy.Persistence.Models;
+
+namespace Shopipy.Shared.Services
+{
+    public interface IProductService
+    {
+        Task<Product> CreateProductAsync(Product product, int businessId);
+        Task<Product?> GetProductByIdAsync(int productId, int businessId);
+        Task<Product?> UpdateProductAsync(int productId, Product product, int businessId);
```
Comment (Buztas) [inline Shopipy.Shared/Services/IProductService.cs:9]: Why are these two nullable?

File: Shopipy.Shared/Services/IProductionService.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using Shopipy.Persistence.Models;
+
+namespace Shopipy.Shared.Services
+{
+    public interface IProductVariationService
+    {
+        Task<ProductVariation> CreateVariationAsync(ProductVariation variation, int productId, int businessId);
+        Task<ProductVariation?> GetVariationByIdAsync(int variationId, int productId, int businessId);
+        Task<ProductVariation?> UpdateVariationAsync(int variationId, ProductVariation variation, int productId, int businessId);
```
Comment (Buztas) [inline Shopipy.Shared/Services/IProductionService.cs:9]: Same here

File: Shopipy.ProductManagement/DependencyInjection.cs (hunk 2 above)
Comment (Buztas) [reply Shopipy.ProductManagement/DependencyInjection.cs]: Also in ApiService program.cs you are not adding product management, how were you even able to call your Http requests without getting DI errors?

File: Shopipy.ProductManagement/Mappings/ProductMappingProfile.cs
Comment (imbieras) [inline Shopipy.ProductManagement/Mappings/ProductMappingProfile.cs:1]: Automapper automatically maps fields with same names.

File: Shopipy.ProductManagement/Controllers/ProductVariationController.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Shopipy.ProductManagement.DTOs;
+using Shopipy.ProductManagement.Services;
+using Shopipy.Persistence.Models;
+using Shopipy.Shared;
+using Shopipy.Shared.Services;
+
+namespace Shopipy.ProductManagement.Controllers
+{
+    [Route("{businessId}/products/{productId}/variations")]
```
Comment (imbieras) [inline Shopipy.ProductManagement/Controllers/ProductVariationController.cs:12]: https://github.com/imbieras/shopipy/blob/0523318744943e2964beba1a74ff2fd0e963b0b3/Shopipy.CategoryManagement/Controllers/CategoryController.cs#L11 or any other Management. Basically this doesn't follow the convention.

File: Shopipy.ProductManagement/DependencyInjection.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+        {
+            categories.AddScoped<IProductService, ProductService>();
```
Comment (imbieras) [inline Shopipy.ProductManagement/DependencyInjection.cs]: What about ProductVariationService :(

File: Shopipy.ProductManagement/Services/ProductService.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+            var product = await _productRepository.GetByConditionAsync(p => p.ProductId == productId && p.BusinessId == businessId);
+            return product;
+        }
+
+        public async Task<Product?> UpdateProductAsync(int productId, Product product, int businessId)
+        {
+            var existingProduct = await _productRepository.GetByConditionAsync(p => p.ProductId == productId && p.BusinessId == businessId);
+            if (existingProduct == null) return null;
+
+            existingProduct.Name = product.Name;
```
Comment (imbieras) [inline Shopipy.ProductManagement/Services/ProductService.cs:30]: Shouldn't be needed.

File: Shopipy.ProductManagement/Services/ProductVariationService.cs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+        public async Task<ProductVariation?> UpdateVariationAsync(int variationId, ProductVariation variation, int productId, int businessId)
+        {
+            var existingVariation = await GetVariationByIdAsync(variationId, productId, businessId);
+            if (existingVariation == null)
+            {
+                return null;
+            }
+
+            existingVariation.Name = variation.Name;
```
Comment (imbieras) [inline Shopipy.ProductManagement/Services/ProductVariationService.cs:41]: Shouldn't be needed.

**Comment by imbieras** [comment]
@Domarkaite use [file scoped namespaces](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/file-scoped-namespaces).

File: Shopipy.ServiceManagement/Properties/launchSettings.json
Comment (imbieras) [inline Shopipy.ServiceManagement/Properties/launchSettings.json:1]: Why is this here, when launching is done for the whole API through Aspire or Docker?

File: Shopipy.UserManagement/Properties/launchSettings.json
Comment (imbieras) [inline Shopipy.UserManagement/Properties/launchSettings.json:1]: Why is this here, when launching is done for the whole API through Aspire or Docker?

File: Shopipy.ProductManagement/Controllers/ProductController.cs
Comment (imbieras) [reply Shopipy.ProductManagement/Controllers/ProductController.cs:1]: Same for the variations, atleast have one method for the variations of a certain product.

Comment (imbieras) [reply Shopipy.ProductManagement/Controllers/ProductController.cs:1]: Also, maybe the GET products could do with pagination. Can grab from user controller

File: Shopipy.Shared/Services/IProductService.cs (hunk 7 above)
Comment (piatrashkakanstantinass) [reply Shopipy.Shared/Services/IProductService.cs:9]: `GetProductByIdAsync` makes sense to return nullable as product may not exist.`UpdateProductAsync` doesn't make sense, also why do you need to specify additional ids in update product? Shouldn't they already be in the `Product` you pass?

File: Shopipy.Shared/Services/IProductionService.cs (hunk 8 above)
Comment (piatrashkakanstantinass) [reply Shopipy.Shared/Services/IProductionService.cs:9]: Same here

File: Shopipy.ProductManagement/Services/ProductVariationService.cs [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+            _variationRepository = variationRepository;
+            _productRepository = productRepository;
+        }
+
+        public async Task<ProductVariation> CreateVariationAsync(ProductVariation variation, int productId, int businessId)
+        {
+            var product = await _productRepository.GetByConditionAsync(p => p.ProductId == productId && p.BusinessId == businessId);
+            if (product == null)
+            {
+                throw new ArgumentException("Product not found for the specified business.");
```
[the remark this answers is not in the document]
Comment (piatrashkakanstantinass) [reply Shopipy.ProductManagement/Services/ProductVariationService.cs:14]: Which is a good thing as we have exception filters that automatically return bad request response with error message

File: Shopipy.ProductManagement/Controllers/ProductController.cs (hunk 1 above)
Comment (Domarkaite) [reply Shopipy.ProductManagement/Controllers/ProductController.cs]: did you mean to include this specific line?

File: Shopipy.ProductManagement/Controllers/ProductController.cs
Comment (piatrashkakanstantinass) [reply Shopipy.ProductManagement/Controllers/ProductController.cs:1]: @imbieras Technically, requirements state that we need pagination in at least one place, we already have it in users endpoint. I agree for this particular case, though, as we can show it in the frontend (unlike users endpoint)

File: Shopipy.ApiService/Program.cs [hunk 14]
```diff
@@ -48,12 +49,13 @@
 }
 
 builder.Services.AddAutoMapper(typeof(UserMappingProfile), typeof(BusinessMappingProfile), typeof(ServiceMappingProfile), 
-    typeof(AppointmentMappingProfile), typeof(CategoryMappingProfile));
+    typeof(AppointmentMappingProfile), typeof(CategoryMappingProfile), typeof(ProductMappingProfile), typeof(ProductVariationMappingProfile));
```
Comment (piatrashkakanstantinass) [inline Shopipy.ApiService/Program.cs:52]: You can keep everything in one mapping profile. Also, I think, you only need to specify 1 mapping profile from every project here as automapper automatically scans for profiles.

File: Shopipy.CategoryManagement/Properties/launchSettings.json [hunk 15]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (piatrashkakanstantinass) [inline Shopipy.CategoryManagement/Properties/launchSettings.json:1]: lauching stuff

File: Shopipy.ProductManagement/Controllers/ProductVariationController.cs [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    [Authorize(Policy = AuthorizationPolicies.RequireBusinessOwnerOrSuperAdmin)]
+    public async Task<ActionResult<ProductVariationResponseDTO>> CreateVariationAsync(int businessId, int productId, ProductVariationRequestDTO dto)
+    {
+        if (!ModelState.IsValid)
+        {
+            return BadRequest(new { error = "Invalid data provided", details = ModelState });
+        }
+
+        var variation = _mapper.Map<ProductVariation>(dto);
+        variation.CreatedAt = DateTime.UtcNow;
```
Comment (piatrashkakanstantinass) [inline Shopipy.ProductManagement/Controllers/ProductVariationController.cs:35]: service concern

File: Shopipy.ProductManagement/DTOs/ProductResponseDTO.cs [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+public class ProductResponseDTO
+{
+    public required int CategoryId { get; set; }
+    public required string Name { get; set; }
+    public required string Description { get; set; }
+    public required decimal BasePrice { get; set; }
+    public required ProductState ProductState { get; set; }
+    public int ProductId { get; set; }
+    public int BusinessId { get; set; }
+    public DateTime CreatedAt { get; set; }
```
Comment (piatrashkakanstantinass) [inline Shopipy.ProductManagement/DTOs/ProductResponseDTO.cs:14]: Add default value of utc now. In this case you will only need to ensure you update `UpdatedAt` inside the service

File: Shopipy.ProductManagement/Controllers/ProductVariationController.cs [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        try
+        {
+            var createdVariation = await _variationService.CreateVariationAsync(variation, productId, businessId);
+            var variationResponseDTO = _mapper.Map<ProductVariationResponseDTO>(createdVariation);
+
+            return CreatedAtAction(nameof(GetVariationByIdAsync),
+                new { businessId, productId, variationId = createdVariation.VariationId },
+                variationResponseDTO);
+        }
+        catch (Exception ex)
```
Comment (piatrashkakanstantinass) [inline Shopipy.ProductManagement/Controllers/ProductVariationController.cs:47]: no catching

File: Shopipy.ProductManagement/Controllers/ProductController.cs (hunk 1 above)
Comment (imbieras) [reply Shopipy.ProductManagement/Controllers/ProductController.cs]: Then we would need to fix all other controllers lololol

File: Shopipy.ProductManagement/DependencyInjection.cs (hunk 2 above)
Comment (Domarkaite) [reply Shopipy.ProductManagement/DependencyInjection.cs]: done

File: Shopipy.ProductManagement/Services/ProductService.cs (hunk 4 above)
Comment (Domarkaite) [reply Shopipy.ProductManagement/Services/ProductService.cs]: done

File: Shopipy.ProductManagement/Services/ProductVariationService.cs (hunk 5 above)
Comment (Domarkaite) [reply Shopipy.ProductManagement/Services/ProductVariationService.cs]: done

File: Shopipy.Shared/Services/IProductService.cs (hunk 7 above)
Comment (Domarkaite) [reply Shopipy.Shared/Services/IProductService.cs]: @piatrashkakanstantinass your comment makes sense, yes. But by the contract, in the request body, we have the product without ids. Same thing with variations.

File: Shopipy.ProductManagement/Services/ProductVariationService.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Domarkaite) [reply Shopipy.ProductManagement/Services/ProductVariationService.cs]: do we

File: Shopipy.Shared/Services/IProductService.cs (hunk 7 above)
Comment (piatrashkakanstantinass) [reply Shopipy.Shared/Services/IProductService.cs]: @Domarkaite don't quite follow, existing product you pass to update should already have id and business id
```cs
            var existingProduct = await _productService.GetProductByIdAsync(productId, businessId);

            if (existingProduct == null) return NotFound();

            _mapper.Map(dto, existingProduct);
            existingProduct.UpdatedAt = DateTime.UtcNow;

            var updatedProduct = await _productService.UpdateProductAsync(productId, existingProduct, businessId);
```

File: Shopipy.ProductManagement/Services/ProductVariationService.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (piatrashkakanstantinass) [reply Shopipy.ProductManagement/Services/ProductVariationService.cs]: If I remember correctly, we do, have you checked?

File: Shopipy.Shared/Services/IProductService.cs (hunk 7 above)
Comment (Domarkaite) [reply Shopipy.Shared/Services/IProductService.cs]: look at the request body in swagger

File: Shopipy.ProductManagement/Controllers/ProductController.cs [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        _productService = productService;
+        _mapper = mapper;
+    }
+
+    [HttpPost]
+    [Authorize(Policy = AuthorizationPolicies.RequireBusinessOwnerOrSuperAdmin)]
+    public async Task<ActionResult<ProductResponseDTO>> CreateProductAsync(ProductRequestDTO dto, int businessId)
+    {
+        if (!ModelState.IsValid)
```
[the remark this answers is not in the document]
Comment (Domarkaite) [reply Shopipy.ProductManagement/Controllers/ProductController.cs:28]: nice

File: Shopipy.ProductManagement/Controllers/ProductVariationController.cs (hunk 18 above)
Comment (Domarkaite) [reply Shopipy.ProductManagement/Controllers/ProductVariationController.cs:47]: wdym

File: Shopipy.ProductManagement/Controllers/ProductController.cs [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    [HttpPut("{productId}")]
+    [Authorize(Policy = AuthorizationPolicies.RequireBusinessOwnerOrSuperAdmin)]
+    public async Task<ActionResult<ProductResponseDTO>> UpdateProduct(int productId, ProductRequestDTO dto, int businessId)
+    {
+        var existingProduct = await _productService.GetProductByIdAsync(productId, businessId);
+
+        if (existingProduct == null) return NotFound();
+
+        _mapper.Map(dto, existingProduct);
+        existingProduct.UpdatedAt = DateTime.UtcNow;
```
Comment (piatrashkakanstantinass) [inline Shopipy.ProductManagement/Controllers/ProductController.cs:64]: Move that to service

File: Shopipy.ProductManagement/Controllers/ProductController.cs [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+using Shopipy.Persistence.Models;
+using Shopipy.ProductManagement.DTOs;
+using Shopipy.Shared;
+using Shopipy.Shared.DTOs;
+using Shopipy.Shared.Services;
+
+namespace Shopipy.ProductManagement.Controllers;
+
+[Route("Product/{businessId}/products")]
+[ApiController]
```
Comment (piatrashkakanstantinass) [inline Shopipy.ProductManagement/Controllers/ProductController.cs:13]: Put `[Authorize]` on the top, everyone will be able to access otherwise (applies to other controllers)

File: Shopipy.ProductManagement/DTOs/ProductResponseDTO.cs (hunk 17 above)
Comment (piatrashkakanstantinass) [inline Shopipy.ProductManagement/DTOs/ProductResponseDTO.cs:14]: UTCNow?

File: Shopipy.ProductManagement/Services/ProductService.cs [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+            products = products.Take(top.Value);
+        }
+
+        return products;
+    }
+
+    public async Task<int> GetProductCountAsync(int businessId)
+    {
+        var products = await _productRepository.GetAllByConditionAsync(p => p.BusinessId == businessId);
+        return products.Count();
```
Comment (piatrashkakanstantinass) [inline Shopipy.ProductManagement/Services/ProductService.cs:37]: extract count to repository as well

File: Shopipy.ServiceManagement/Properties/launchSettings.json [hunk 23]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (piatrashkakanstantinass) [inline Shopipy.ServiceManagement/Properties/launchSettings.json:1]: wow, some more launching

File: Shopipy.UserManagement/Properties/launchSettings.json [hunk 24]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (piatrashkakanstantinass) [inline Shopipy.UserManagement/Properties/launchSettings.json:1]: crazy

**Review by piatrashkakanstantinass — APPROVED** [review]
